plugins/inputs/fritzbox/fritzbox.go

```diff
@@ -0,0 +1,420 @@
+		if err != nil {
+			return err
+		}
+		client := tr064.NewClient(parsedUrl)
+		client.Debug = plugin.Log.Level().Includes(telegraf.Debug)
+		client.Timeout = time.Duration(plugin.Timeout)
+		tlsConfig, err := plugin.TLSConfig()
+		if err != nil {
+			return err
+		}
+		client.TlsConfig = tlsConfig
+		plugin.deviceClients = append(plugin.deviceClients, client)
+	}
+	return nil
+}
+
+func (plugin *Fritzbox) initServiceHandlers() {
+	if choice.Contains("device", plugin.Collect) {
+		plugin.serviceHandlers[deviceinfo.ServiceShortType] = plugin.gatherDeviceInfo
+	}
+	if choice.Contains("wan", plugin.Collect) {
+		plugin.serviceHandlers[wancommonifconfig.ServiceShortType] = plugin.gatherWanInfo
+	}
+	if choice.Contains("ppp", plugin.Collect) {
+		plugin.serviceHandlers[wanpppconn.ServiceShortType] = plugin.gatherPppInfo
+	}
+	if choice.Contains("dsl", plugin.Collect) {
+		plugin.serviceHandlers[wandslifconfig.ServiceShortType] = plugin.gatherDslInfo
+	}
+	if choice.Contains("wlan", plugin.Collect) {
+		plugin.serviceHandlers[wlanconfig.ServiceShortType] = plugin.gatherWlanInfo
+	}
+	if choice.Contains("hosts", plugin.Collect) {
+		plugin.serviceHandlers[hosts.ServiceShortType] = plugin.gatherHostsInfo
+	}
+}
+
+func (plugin *Fritzbox) Gather(acc telegraf.Accumulator) error {
+	var waitComplete sync.WaitGroup
+	for _, deviceClient := range plugin.deviceClients {
+		waitComplete.Add(1)
+		go func() {
+			defer waitComplete.Done()
+			plugin.gatherDevice(acc, deviceClient)
+		}()
+	}
+	waitComplete.Wait()
+	return nil
+}
+
+func (plugin *Fritzbox) gatherDevice(acc telegraf.Accumulator, deviceClient *tr064.Client) {
+	plugin.Log.Debugf("Querying %s", deviceClient.DeviceUrl.Redacted())
+	services, err := deviceClient.Services(tr064.DefaultServiceSpec)
+	if err != nil {
+		acc.AddError(err)
+		return
+	}
+	for _, service := range services {
+		serviceHandler := plugin.serviceHandlers[service.ShortType()]
+		if serviceHandler == nil {
+			continue
+		}
+		err := serviceHandler(acc, deviceClient, service)
+		if err != nil {
+			acc.AddError(err)
+			return
+		}
+	}
+}
+
+type serviceHandlerFunc func(telegraf.Accumulator, *tr064.Client, tr064.ServiceDescriptor) error
+
+func (plugin *Fritzbox) gatherDeviceInfo(acc telegraf.Accumulator, deviceClient *tr064.Client, service tr064.ServiceDescriptor) error {
+	serviceClient := deviceinfo.ServiceClient{
+		TR064Client: deviceClient,
+		Service:     service,
+	}
+	info := &deviceinfo.GetInfoResponse{}
+	err := serviceClient.GetInfo(info)
+	if err != nil {
+		return err
+	}
+	tags := make(map[string]string)
+	tags["source"] = serviceClient.TR064Client.DeviceUrl.Hostname()
+	tags["service"] = serviceClient.Service.ShortId()
+	fields := make(map[string]interface{})
+	fields["uptime"] = info.NewUpTime
+	fields["model_name"] = info.NewModelName
+	fields["serial_number"] = info.NewSerialNumber
+	fields["hardware_version"] = info.NewHardwareVersion
+	fields["software_version"] = info.NewSoftwareVersion
```

Comment:
Please declare non-conditional maps like
```suggestion
	tags := map[string]string{
		"source": serviceClient.TR064Client.DeviceUrl.Hostname(),
		"service": serviceClient.Service.ShortId(),
	}
	fields := map[string]interface{}{
		"uptime"] = info.NewUpTime,
		"model_name": info.NewModelName,
		"serial_number": info.NewSerialNumber,
		"hardware_version": info.NewHardwareVersion,
		"software_version": info.NewSoftwareVersion,
	}
```
Same for all spots below.

Reply:
Adapted via 1ada7a8c36211e8f74afa86ed9638c45083686bb